hugo/content/docs/recipes/utilities/caches.md

```diff
@@ -0,0 +1,223 @@
+
+* the document cache saves key-value-pairs of given types `K` and `V` for each document. If the document gets changed or deleted the cache gets cleared automatically for the single files
+* the workspace cache also save key-value-pairs of given types `K` and `V` but gets cleared entirely when something in the workspace gets changed.
+
+Besides those specific caches, Langium also provides
+
+* a simple cache that can be used for any kind of key-value-data
+* a context cache that stores a simple cache for each context object. The document cache and workspace cache are implemented using the context cache.
+
+## How to use it?
+
+Here we will use the `HelloWorld` example from the learning section. Let's keep it simple and just list persons in the document, that come from a comic book.
+
+We will have a computation for each person that determines from which publisher it comes from.
+
+### Add a database
+
+Let's build a "publisher inferer service". First let's create a small database of known publishers and known persons:
+
+```typescript
+type KnownPublisher = 'DC'|'Marvel'|'Egmont';
+const KnownPersonNames: Record<KnownPublisher, string[]> = {
+    DC: ['Superman', 'Batman', 'Aquaman', 'Wonderwoman', 'Flash'],
+    Marvel: ['Spiderman', 'Wolverine', 'Deadpool'],
+    Egmont: ['Asterix', 'Obelix']
+};
+```
+
+### Define the computation service
+
+For our service we define an interface:
+
+```typescript
+export interface InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined;
+}
+```
+
+Now we implement the service:
+
+```typescript
+class UncachedInferPublisherService implements InferPublisherService {
+    inferPublisher(person: Person): KnownPublisher|undefined {
+        for (const [publisher, persons] of Object.entries(KnownPersonNames)) {
+            if(persons.includes(person.name)) {
+                return publisher as KnownPublisher;
+            }
+        }
+        return undefined;
+    }
+}
+```
+
+### Add a cache
+
+Now we want to cache the results of the `inferPublisher` method. We can use the `DocumentCache` for this. We will reuse the uncached service as base class and override the `inferPublisher` method:
+
+```typescript
+export class CachedInferPublisherService extends UncachedInferPublisherService {
+    private readonly cache: DocumentCache<Person, KnownPublisher|undefined>;
+    constructor(services: HelloWorldServices) {
+        super();
+        this.cache = new DocumentCache(services.shared);
+    }
+    override inferPublisher(person: Person): KnownPublisher|undefined {
+        const documentUri = AstUtils.getDocument(person).uri;
+        if(this.cache.has(documentUri, person)) {
+           return this.cache.get(documentUri, person)!;
+        }
+        const publisher = super.inferPublisher(person);
+        this.cache.set(documentUri, person, publisher);
+        return publisher;
```

Comment:
Demonstrating the use of `ContextCache.get(contextKey: Context, key: Key, provider: () => Value): Value` might also be useful:

```ts
const documentUri = AstUtils.getDocument(person).uri;
return this.cache.get(documentUri, person, () => super.inferPublisher(person));
```

Reply:
Uh! Very nice! I was not aware of, thank you!